js/server/modules/@arangodb/foxx/manager.js

```diff
@@ -620,15 +633,50 @@ function _buildServiceFromFile (tempServicePath, tempBundlePath, filePath) {
   fs.copyFile(filePath, tempBundlePath);
 }
 
-function _buildServiceBundleFromScript (tempServicePath, tempBundlePath, jsBuffer) {
-  const manifest = JSON.stringify({main: 'index.js'}, null, 4);
+function _buildServiceBundleFromScript(tempServicePath, tempBundlePath, jsBuffer) {
+  const manifest = JSON.stringify({ main: 'index.js' }, null, 4);
   fs.makeDirectoryRecursive(tempServicePath);
   fs.writeFileSync(path.join(tempServicePath, 'index.js'), jsBuffer);
   fs.writeFileSync(path.join(tempServicePath, 'manifest.json'), manifest);
   utils.zipDirectory(tempServicePath, tempBundlePath);
 }
 
-function _deleteServiceFromPath (mount, force) {
+function _deleteServiceFromPath(mount, force) {
+}
+
+function _checkServiceModificationKeyspace() {
+  if (!global.KEYSPACE_EXISTS('foxx-service-lock')) {
+    global.KEYSPACE_CREATE('foxx-service-lock', 1, true);
+    _setServiceModificationLock(false);
+  }
+}
+
+function _getServiceModificationLock() {
+  _checkServiceModificationKeyspace();
+  const boolRes = global.KEY_GET('foxx-service-lock', 'is-service-locked') || false;
+  return boolRes;
+}
+
+function _setServiceModificationLock(toLock) {
+  if (typeof toLock !== 'boolean') {
+    return;
+  }
+  global.KEY_SET('foxx-service-lock', 'is-service-locked', toLock);
+}
+
+function _checkServiceModificationLock() {
+  try {
+    for (let count = 0; count < 10; count++) {
+      if (_getServiceModificationLock()) {
+        sleep(1);
+      } else {
+        return _getServiceModificationLock();
+      }
+    }
+  } catch (ignore) {
+  }
+
+  return _getServiceModificationLock();
```

Comment:
That you're not trusting the result of that method to be consistent between the "if" and "return" means we shouldn't trust the return value of this method either. For good reason.

Checking-and-acquiring a lock should probably be atomic. This currently has a race condition. Shouldn't this be handled by the agency somehow?